### PR TITLE
Made Body::containsPoint() consistent in surface points inclusion. 

### PR DESCRIPTION
**Pull request by peci1**

### CODE REVIEW

**Comment by peci1** [comment]
See https://github.com/ros-planning/geometric_shapes/issues/91 for context.

File: README.md [hunk 1]
```diff
@@ -13,12 +13,14 @@ Supported shapes:
...
+Note: There was a recent change in point containment computations - now all bodies are expected to also contain their surface points. [The old behavior was inconsistent](https://github.com/ros-planning/geometric_shapes/issues/91). Rounding errors may still yield surprising results, though.
+
 Note: [shape_tools](https://github.com/ros-planning/shape_tools) package was recently merged into this package
 
 ## Build Status
 
 Travis CI: [![Build Status](https://travis-ci.org/ros-planning/geometric_shapes.svg?branch=indigo-devel)](https://travis-ci.org/ros-planning/geometric_shapes)
 
-Devel Job: [![Build Status](http://build.ros.org/buildStatus/icon?job=Jbin_uT64__geometric_shapes__ubuntu_trusty_amd64__binary)](http://build.ros.org/view/Ibin_uT64/job/Jbin_uT64__geometric_shapes__ubuntu_trusty_amd64__binary/)
+Devel Job: [![Build Status](http://build.ros.org/buildStatus/icon?job=Ibin_uT64__geometric_shapes__ubuntu_trusty_amd64__binary)](http://build.ros.org/view/Ibin_uT64/job/Ibin_uT64__geometric_shapes__ubuntu_trusty_amd64__binary/)
```
Comment (rhaschke) [inline README.md:24]: Why did you switched these badges from Jade to Indigo?
They should actually be Melodic...

File: test/test_point_inclusion.cpp [hunk 2]
```diff
@@ -66,7 +66,14 @@ TEST(SpherePointContainment, Basic)
...
 
     // near two-axis maximum
-    const double sq2 = sqrt(2) / 2;
+    // HACK: sqrt(2) / 2 is a problem due to rounding errors since the distance
+    //       is computed as 1.0000000000000002 . In such case we subtract an
+    //       epsilon.
+    double sq2 = sqrt(2) / 2;
+    if (sq2 * sq2 + sq2 * sq2 > 1)
+      sq2 -= std::numeric_limits<double>::epsilon();
+    ASSERT_LE(sq2 * sq2 + sq2 * sq2, 1.0);
```
Comment (rhaschke) [inline test/test_point_inclusion.cpp]: [`EXPECT_DOUBLE_EQ`](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#floating-point-comparison) comes to rescue!

File: test/test_point_inclusion.cpp [hunk 3]
```diff
@@ -359,9 +366,17 @@ TEST(CylinderPointContainment, Basic)
     EXPECT_FALSE(cylinder.containsPoint(Eigen::Vector3d(0.00, 0.00, 2.01)));
 
     // near two-axis maximum
+    // HACK: sqrt(2) / 2 is a problem due to rounding errors since the distance
+    //       is computed as 1.0000000000000002 . In such case we subtract an
+    //       epsilon. For the case when 1+sq2 is used, this does not occur.
     const double sq2 = sqrt(2) / 2;
+    double sq2_eps = sq2;
```
Comment (rhaschke) [inline test/test_point_inclusion.cpp]: Same here of course.

File: test/test_point_inclusion.cpp [hunk 4]
```diff
@@ -359,9 +366,17 @@ TEST(CylinderPointContainment, Basic)
...
+    if (sq2_eps * sq2_eps + sq2_eps * sq2_eps > 1)
+      sq2_eps -= std::numeric_limits<double>::epsilon();
+    ASSERT_LE(sq2_eps * sq2_eps + sq2_eps * sq2_eps, 1.0);
+
     EXPECT_TRUE( cylinder.containsPoint(Eigen::Vector3d(0.70, 0.70, 0.00)));
-    EXPECT_SURF( cylinder.containsPoint(Eigen::Vector3d(sq2 , sq2 , 0.00)));
+    EXPECT_SURF(cylinder.containsPoint(Eigen::Vector3d(sq2_eps , sq2_eps , 0)));
```
Comment (rhaschke) [inline test/test_point_inclusion.cpp]: Please revert these changes. They shouldn't be necessary anymore with `EXPECT_DOUBLE_EQ`.

**Review by rhaschke — CHANGES_REQUESTED** [review]
Thanks a lot. Some minor comments remain.

File: README.md (hunk 1 above)
Comment (peci1) [reply README.md:24]: This is targeted at indigo-devel... The badges as were here were broken: `view/Ibin_uT64/job/Jbin_uT64__geometric_shapes` (notice `Ibin` first and `Jbin` second).

File: test/test_point_inclusion.cpp (hunk 2 above)
Comment (peci1) [reply test/test_point_inclusion.cpp]: Nope: EXPECT_DOUBLE_EQ doesn't help with this:  `EXPECT_SURF( sphere.containsPoint(Eigen::Vector3d(sq2 , sq2 , 0.00)));`

File: README.md (hunk 1 above)
Comment (rhaschke) [reply README.md:24]: > This is targeted at indigo-devel

Uuh. Didn't noticed. @v4hn @davetcoleman Should we allow this fix for Indigo?
It will change (the inconsistent) behaviour to be consistent. Although not a pure bug fix, it's a fix.

> The badges were broken

I see. Can you please re-organize your PR branch into three distinct commits:
- fixing the badges and
- adding the failing unit test
- fixing the inconsistency, making the unittest happy

This will facilitate cherry-picking to the other branches.

Comment (peci1) [reply README.md:24]: Sure, I can reorganize. Do you know if force-pushing is okay when the branch is involved in a PR?

Comment (peci1) [reply README.md:24]: Seems it worked...

File: test/test_point_inclusion.cpp (hunk 2 above)
Comment (rhaschke) [reply test/test_point_inclusion.cpp]: OK. I see. As you use this "hacky" code twice, could you please make a small function from it.
```cpp
// split length of 1 into the largest number elem, such that sqrt(elem^2 + elem^2) <= 1
double elem = largest_component_for_length(1.0)
```

File: README.md (hunk 1 above)
Comment (rhaschke) [reply README.md:24]: Yes, force-pushing is frequently used to cleanup PR branches.

**Review by rhaschke — CHANGES_REQUESTED** [review]
Looks good to me. One final request: Could you please squash the last two commits (bug fix + refactoring) into one. Then I propose a merge-commit to keep the clean history of this PR branch.

**Comment by peci1** [comment]
I squashed and force-pushed. I didn't quite get the part about merge commit - there seems to be nothing to merge... Or did you mean merge instead of force-push?

File: test/test_point_inclusion.cpp [hunk 5]
```diff
@@ -360,8 +373,10 @@ TEST(CylinderPointContainment, Basic)
 
     // near two-axis maximum
     const double sq2 = sqrt(2) / 2;
```
Comment (rhaschke) [inline test/test_point_inclusion.cpp:375]: Looks like sq2 is only used in a single location. Do you mind to use sq2_eps for this location too, thus allowing to rename s2_eps back to sq2. This way, the bugfix wouldn't require too many changes for the test.

**Review by rhaschke — CHANGES_REQUESTED** [review]
One more nitpick.

**Comment by rhaschke** [comment]
> I didn't quite get the part about merge commit - there seems to be nothing to merge.

To merge this PR branch into its parent, github provides several options: explicit merge commit, rebase, or squash-rebase. To keep the history of your 3 commits (thanks a lot cleaning them up), I suggested the merge commit.

File: test/test_point_inclusion.cpp (hunk 5 above)
Comment (peci1) [reply test/test_point_inclusion.cpp:375]: It is actually used at 2 locations: line 400 and 406. I wanted to use this "clean" value where it works, and only use sq2_eps at places which are really affected by IEEE 754 rounding errors. If you just don't like the name sq2_eps, I can try finding a new one :)

**Comment by peci1** [comment]
Hmm, is this something new on github? I don't see any such setting in the PR options from my side. It looks to me like options for the one who is finally merging this PR, which is not me (I don't have commit access). Or do I miss a point? Should I create an explicit merge commit in my repo and then send a PR from my indigo-devel to upstream?

**Comment by rhaschke** [comment]
Hope [this](https://github.com/ros-planning/moveit.ros.org/blob/fdbb1ee0d90321de5e538adba78a75a3ac2a95f4/documentation/contributing/pullrequests/index.markdown#github-merge-policies) helps to explain the different policies.
You are right, only maintainers with push access see the mentioned merge options. 
No other action than https://github.com/ros-planning/geometric_shapes/pull/90#discussion_r251234245 is required for you.

**Comment by peci1** [comment]
Ok, got it. There's still an unanswered question in https://github.com/ros-planning/geometric_shapes/pull/90#discussion_r251234245 .

File: test/test_point_inclusion.cpp (hunk 5 above)
Comment (rhaschke) [reply test/test_point_inclusion.cpp:375]: Yeah, I don't like the name very much. But it don't have a better idea either. I'm not so strong on names ;-)

**Comment by rhaschke** [comment]
> There's still an unanswered question in #90 (comment) .

Sorry. Didn't noticed this one. Answered now.

File: test/test_point_inclusion.cpp (hunk 5 above)
Comment (peci1) [reply test/test_point_inclusion.cpp:375]: I renamed `sq2_eps` to `sq2e`. Also, to keep consistency throughout the code, I renamed `sq2` to `sq2e` in the sphere test.

File: test/test_point_inclusion.cpp [hunk 6]
```diff
@@ -360,8 +373,10 @@ TEST(CylinderPointContainment, Basic)
...
+    const double sq2e = largestComponentForLength2D(1);
+
     EXPECT_TRUE( cylinder.containsPoint(Eigen::Vector3d(0.70, 0.70, 0.00)));
-    EXPECT_SURF( cylinder.containsPoint(Eigen::Vector3d(sq2 , sq2 , 0.00)));
+    EXPECT_SURF(cylinder.containsPoint(Eigen::Vector3d(sq2e , sq2e , 0)));
```
Comment (rhaschke) [inline test/test_point_inclusion.cpp]: Better. Could you also restore the previous fixed number format: `0 -> 0.00` and clean spaces, commas, etc. 
Your original code looked better in this regard. :wink:

Comment (peci1) [reply test/test_point_inclusion.cpp]: Done.

**Review by rhaschke — APPROVED** [review]
Great! Thanks for your patience.

File: README.md (hunk 1 above)
Comment (davetcoleman) [reply README.md:24]: The breakage risk to Indigo doesn't seem worth it to me. @peci1 could you start working off of the melodic-devel branch instead? are you still on indigo for everything?

**Review by davetcoleman — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (peci1) [reply README.md:24]: When I started this PR, we were fully on indigo, now we're transitioning to melodic. Sure, I can rebase this PR on top of melodic devel. Kinetic remains a question, though...

Comment (peci1) [reply README.md:24]: But now I'm a bit confused as I see you also approved this PR...

Comment (peci1) [reply README.md:24]: @davetcoleman ?

Comment (peci1) [reply README.md:24]: @davetcoleman See #97 - I've retargeted this PR to melodic.

**Comment by rhaschke** [comment]
Closing this in favor of #97.
